### PR TITLE
Sky130 Simulations + Auto-Include

### LINKED ISSUE #109: Auto-Generated PDK Device Mapping Tables
Feature: automagically generate documentation tables along the lines of: 

---

### Mos Device Mapping 

| MosType   | MosFamily |    MosVth  |  Device Name |
|-------- | -------- | -------- | -------- | 
| NMOS   | CORE |    STD  |  mypdk_nch |
| PMOS   | CORE |    STD  |  mypdk_pch |
| NMOS   | CORE |    LOW  |  mypdk_nch_lvt |
| PMOS   | CORE |    LOW  |  mypdk_pch_lvt |
| NMOS   | IO |    STD  |  mypdk_nch_io |
| PMOS   | IO |    STD  |  mypdk_pch_io |

(etc)

Maybe this could add some more detail such as valid or default parameter values too. But primarily, tabulate the mappings.

### PR DESCRIPTION
**Pull request by ThomasPluck**
This a twin PR with https://github.com/Vlsir/Vlsir/pull/48

## Description
- [x] Define tests for simulations of all SKY130 components
- [x] New PHYSICAL/IDEAL method for netlisting `ExternalModules`
- [x] Successfully netlist and simulate all components
- [x] Provide auto-includes to minimize ngspice import time
~~- [ ] Improve Github Action to also include simulators in test environment~~
~~- [ ] Resolve #109~~

## Related PRs and issues
- https://github.com/Vlsir/Vlsir/pull/48
- #95
- #107 
- #110

## Additional notes

- "Transistory" components work fine, including varicaps, "generic" and precision resistors and all transistors.
- The broken but fixable components fall into one of two camps:
  - Need their netlist prefix changing in `Vlsirtools`
  - Are poorly documented, either not being mentioned or being misnamed.
- Some ESD transistors, RF diodes, many of the VPPs simply do not work as intended even in other netlisting programs.

### CODE REVIEW

File: hdl21/external_module.py [hunk 1]
```diff
@@ -39,7 +38,7 @@ class ExternalModule:
     paramtype: Type = HasNoParams  # Parameter-type `paramclass`
     desc: Optional[str] = None  # Description
     domain: Optional[str] = None  # Domain name, for references upon export
-    prefix: Optional[str] = None # Prefix that overrides default netlist naming
+    devicetype: Optional[str] = None # Device type, takes its cue from primitives.py
```
Comment (dan-fritchman) [inline hdl21/external_module.py:41]: I don't wanna add this.

Comment (ThomasPluck) [reply hdl21/external_module.py:41]: Yeah, that's fair, this is definitely not the HDL21 way of doing things. So I guess we could do the following:

1) Define a new device enum in primitives that basically just replicates VLSIR's SpicePrefix
2) Pass this enum to the ExternalModule's "property"
3) Make `to_proto` sift through properties looking for instances of the device enum
4) Tack on the enum's value onto the pkg protobuf as a param and let patched VLSIR work it's magic.

I'm just happy that I got those resistors to work! What do you think?

Comment (ThomasPluck) [reply hdl21/external_module.py]: RE: Latest Commit

- Added props to ExternalModule class to reflect Module class
- Persisted Sim simulator context into Tb.props
- Modified "to_proto" to look for SpicePrefix objects in Inst.module.props matched with simulator context in tops.module.props
- If no simulator is specified by the instance, default to SpicePrefix.SUBCKT
- If a simulator is specified but set to "NotImplemented" in the ExternalModule, throws a NotImplemented error. If simulator isn't described by ExternalModule, error will be thrown when program looks for said simulator key.
- This is then handled as before by passing the *value* of the chosen SpicePrefix to param inside the protobuf which is then parsed by vlsirtools into the correct spice prefix needed to run a component.

**Comment by ThomasPluck** [comment]
Expanded the documentation with the help of ChatGPT. Can't resolve testing environment issues due to data cap issues on my end. I am now opening this PR to official review.

File: pdks/Sky130/sky130/pdk_data.py [hunk 2]
```diff
@@ -0,0 +1,1078 @@
...
+    modname: str,
+    numterminals: int,
+    params: h.Param,
+    props: Dict = {
+        "ngspice": SpicePrefix.SUBCKT,
+        "xyce": SpicePrefix.SUBCKT,
+        "spectre": NotImplemented,
+        "deps": Dict,
+    },
+    deps: Dict = {},
```
Comment (growly) [inline pdks/Sky130/sky130/pdk_data.py:363]: 'override_deps' or 'explicit_deps'

File: hdl21/proto/to_proto.py [hunk 3]
```diff
@@ -174,7 +175,7 @@ def export_external_module(self, emod: ExternalModule) -> vckt.ExternalModule:
         self.pkg.ext_modules.append(pmod)
         return pmod
 
-    def export_instance(self, inst: Instance) -> vckt.Instance:
+    def export_instance(self, module: Module, inst: Instance) -> vckt.Instance:
```
Comment (growly) [inline hdl21/proto/to_proto.py:178]: is `module` available through `inst.of`?

**Comment by ThomasPluck** [comment]
Trying to use the library myself, encountering a few weird issues:

1) modules must now be imported as `from sky130.pdk_logic import modules as s`
2) 'compilation' certainly doesn't work as intended and must be checked again

**Pull request merged**